File: server/ops/environment/baseline.js
// "Unusually high, or climbing without a clear cause" — defined honestly,
// not asserted. A rolling median over real recent history (env_samples),
// plus a SUSTAINED-duration requirement so one genuine but brief spike
// never fires a finding — the same "don't fire on a single blip" discipline
// heartbeat/commitments-source.js's own notifiedApproaching/notifiedOverdue
// flags already established for a different kind of noisy signal.
//
// Leaf module: imports only env-store.js (itself a leaf) — safe for
// server/tools/check_environment.js.

import { listRecentSamples } from './env-store.js';

// A window long enough to establish a real baseline without reacting to
// this build's own startup transient (the very first sample has no CPU
// delta at all — see sampler.js).
const BASELINE_WINDOW_MS = 60 * 60 * 1000; // 1 hour
// How far above the baseline median counts as "unusually high" at all.
const CPU_ELEVATED_THRESHOLD_PCT = 1.5; // 1.5x the recent median
const MEM_LOW_FREE_THRESHOLD_PCT = 10; // below 10% free is elevated concern regardless of history
// A spike must hold for this long, continuously, before it's a finding —
// not a single 30s sample.
const SUSTAINED_DURATION_MS = 5 * 60 * 1000; // 5 minutes

function median(values) {
  if (!values.length) return null;
  const sorted = [...values].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  return sorted.length % 2 ? sorted[mid] : (sorted[mid - 1] + sorted[mid]) / 2;
}

/**
 * Checks whether CPU (or low free memory) has been sustained-elevated for
 * at least SUSTAINED_DURATION_MS, judged against this machine's OWN recent
 * median rather than a fixed number that would mean something different on
 * every machine this ever runs on. Returns null when nothing is unusual, or
 * a plain-language finding when it is. Never fires on fewer than a handful
 * of real samples — "unusually high" is meaningless without a real
 * baseline to compare against.
 */
export function checkForAnomaly() {
  const since = new Date(Date.now() - BASELINE_WINDOW_MS).toISOString();
  const samples = listRecentSamples({ sinceIso: since, limit: 1000 });
  const withCpu = samples.filter((s) => typeof s.cpuPct === 'number');
  if (withCpu.length < 5) return null; // not enough real history yet

  const cpuMedian = median(withCpu.map((s) => s.cpuPct));
  const cpuThreshold = cpuMedian * CPU_ELEVATED_THRESHOLD_PCT;

  // Walk backward from the most recent sample; a sustained window means
  // every sample within SUSTAINED_DURATION_MS of "now" was elevated, with
  // no gap that would mean it actually settled and rose again separately.
  const sortedDesc = [...withCpu].sort((a, b) => new Date(b.ts) - new Date(a.ts));
  const sustainedCutoff = Date.now() - SUSTAINED_DURATION_MS;
  const recentWindow = sortedDesc.filter((s) => new Date(s.ts).getTime() >= sustainedCutoff);
  const oldestInWindow = recentWindow[recentWindow.length - 1];
  const windowSpansFullDuration = oldestInWindow && new Date(oldestInWindow.ts).getTime() <= sustainedCutoff + 60_000; // real coverage, not a gap
  const allElevated = recentWindow.length > 0 && recentWindow.every((s) => s.cpuPct >= cpuThreshold);

  if (allElevated && windowSpansFullDuration && cpuThreshold > 5) {
    // cpuThreshold > 5 guards a near-idle machine (median ~1%, threshold
    // ~1.5%) from reporting trivial noise as "unusually high."
    return {
      summary: `CPU load has been sustained-elevated for the last ${Math.round(SUSTAINED_DURATION_MS / 60000)} minutes — currently around ${Math.round(recentWindow[0].cpuPct)}%, versus a recent typical median of ${Math.round(cpuMedian)}%.`,
      detail: JSON.stringify({ kind: 'cpu', current: recentWindow[0].cpuPct, median: cpuMedian, sustainedMs: SUSTAINED_DURATION_MS }),
    };
  }

  const latest = sortedDesc[0];
  if (latest && typeof latest.memFreePct === 'number' && latest.memFreePct < MEM_LOW_FREE_THRESHOLD_PCT) {
    const memRecentWindow = sortedDesc.filter((s) => new Date(s.ts).getTime() >= sustainedCutoff && typeof s.memFreePct === 'number');
    const memSustained = memRecentWindow.length > 0 && memRecentWindow.every((s) => s.memFreePct < MEM_LOW_FREE_THRESHOLD_PCT);
    if (memSustained) {
      return {
        summary: `Free memory has been below ${MEM_LOW_FREE_THRESHOLD_PCT}% for the last ${Math.round(SUSTAINED_DURATION_MS / 60000)} minutes — currently around ${Math.round(latest.memFreePct)}% free.`,
        detail: JSON.stringify({ kind: 'memory', current: latest.memFreePct, sustainedMs: SUSTAINED_DURATION_MS }),
      };
    }
  }

  return null;
}

File: backend/jarvis/ops/environment/baseline.py
"""'Unusually high, or climbing' — defined against this machine, and sustained.

Two decisions do all the work here, and both exist because the alternative
produces a finding nobody can act on:

* **Compared against this machine's own recent median**, never a fixed
  percentage. "80% CPU" means something different on a laptop that idles at 40%
  and one that idles at 3%.
* **Sustained for a real duration**, never a single sample. A build, a video
  call or opening a large file spikes the machine constantly; a notification for
  each is noise, and noise is how a real one gets ignored.

Pure: it is handed samples, or reads them, and returns a finding or None. No
model call, no interruption decision — that judgment belongs to the heartbeat.
"""

from __future__ import annotations

import json
from datetime import datetime, timedelta, timezone
from typing import Any

from ...jscompat import to_iso_z
from . import store

BASELINE_WINDOW = timedelta(hours=1)
#: How far above this machine's own median counts as elevated at all.
ELEVATED_MULTIPLIER = 1.5
#: Below this share of memory free is a concern whatever the history says.
LOW_FREE_MEMORY_PCT = 10.0
SUSTAINED = timedelta(minutes=5)
#: A near-idle machine has a tiny median, so 1.5x of it is still nothing. Without
#: this, a median of 1% would report 1.6% as "unusually high".
MINIMUM_INTERESTING_CPU_PCT = 5.0
#: "Unusually high" needs something to be unusual against.
MINIMUM_SAMPLES = 5


def _median(values: list[float]) -> float:
    ordered = sorted(values)
    middle = len(ordered) // 2
    if len(ordered) % 2:
        return ordered[middle]
    return (ordered[middle - 1] + ordered[middle]) / 2


def _at(sample: dict[str, Any]) -> datetime:
    return datetime.fromisoformat(str(sample["ts"]).replace("Z", "+00:00"))


def _covers_the_full_duration(window: list[dict[str, Any]], cutoff: datetime) -> bool:
    """The window must actually REACH back to the cutoff. A run of three elevated
    samples in the last ninety seconds is not five sustained minutes — and
    without this check it reads as though it were."""
    if not window:
        return False
    oldest = _at(window[-1])
    return oldest <= cutoff + timedelta(minutes=1)


def check_for_anomaly(samples: list[dict[str, Any]] | None = None,
                      now: datetime | None = None) -> dict[str, Any] | None:
    now = now or datetime.now(timezone.utc)
    if samples is None:
        samples = store.list_recent(since_iso=to_iso_z(now - BASELINE_WINDOW), limit=1000)

    with_cpu = [s for s in samples if isinstance(s.get("cpuPct"), (int, float))]
    if len(with_cpu) < MINIMUM_SAMPLES:
        return None                      # no real baseline yet; silence is correct

    newest_first = sorted(with_cpu, key=_at, reverse=True)
    cutoff = now - SUSTAINED
    window = [s for s in newest_first if _at(s) >= cutoff]

    median = _median([float(s["cpuPct"]) for s in with_cpu])
    threshold = median * ELEVATED_MULTIPLIER

    if (threshold > MINIMUM_INTERESTING_CPU_PCT and window
            and all(float(s["cpuPct"]) >= threshold for s in window)
            and _covers_the_full_duration(window, cutoff)):
        minutes = int(SUSTAINED.total_seconds() // 60)
        return {
            "summary": (f"CPU load has stayed high for the last {minutes} minutes — "
                        f"around {round(float(window[0]['cpuPct']))}%, against a recent "
                        f"typical {round(median)}% on this machine."),
            "detail": json.dumps({"kind": "cpu", "current": window[0]["cpuPct"],
                                  "median": median,
                                  "sustainedMinutes": minutes}),
        }

    with_memory = [s for s in newest_first
                   if isinstance(s.get("memFreePct"), (int, float)) and _at(s) >= cutoff]
    if (with_memory and all(float(s["memFreePct"]) < LOW_FREE_MEMORY_PCT for s in with_memory)
            and _covers_the_full_duration(with_memory, cutoff)):
        minutes = int(SUSTAINED.total_seconds() // 60)
        return {
            "summary": (f"Free memory has been under {round(LOW_FREE_MEMORY_PCT)}% for the "
                        f"last {minutes} minutes — around "
                        f"{round(float(with_memory[0]['memFreePct']))}% free now."),
            "detail": json.dumps({"kind": "memory", "current": with_memory[0]["memFreePct"],
                                  "sustainedMinutes": minutes}),
        }
    return None
